crates/host/src/guest.rs

```diff
@@ -114,28 +179,34 @@ fn call_inner(
     call: &str,
     payload: SerializedBytes,
 ) -> Result<SerializedBytes, WasmError> {
-    // @TODO this is insecure because it leaks the payload and relies on the guest to consume it
-    // with host_args!()
-    // if the guest never consumes this ptr then the payload stays leaked so we want to fix/guard
-    // against that by dropping the payload bytes if the guest never uses them
-    let host_allocation_ptr: AllocationPtr = payload.into();
-
-    // this requires that the guest exported function being called knows what to do with a
-    // host allocation pointer
-    let guest_allocation_ptr: RemotePtr = match instance
+    // get a pre-allocated guest pointer to write the input into
+    let guest_input_ptr: GuestPtr = match instance
         .call(
-            call,
-            &[Value::I64(host_allocation_ptr.as_remote_ptr().try_into()?)],
+            "__allocate",
+            &[Value::I32(payload.bytes().len().try_into()?)],
         )
-        .expect("call error")[0]
+        .map_err(|e| WasmError::CallError(format!("{:?}", e)))?[0]
+    {
+        Value::I32(i) => i as GuestPtr,
```

Comment:
ahh, maybe this makes my previous two comments even more irrelevant... unless you wanted to switch/match on WasmSize somehow to capture the correct wasmer value type..

Reply:
@neonphog yeah this is an example of where rust and wasm don't _quite_ match and there is a bit of friction, the goal is just to keep it to a minimum so this bit is ok i think